tests/test_drvsupport.py

```diff
@@ -2,13 +2,363 @@
+
+        with fiona.open(path) as c:
+            assert c.driver == driver
+            assert len([f for f in c]) == 2
+
+
+@pytest.mark.parametrize('driver', [driver for driver in driver_mode_mingdal['a'].keys()
+                                    if driver not in blacklist_append_drivers
+                                    and driver in supported_drivers])
+def test_append_does_not_work_when_gdal_smaller_mingdal(tmpdir, driver):
+    """ Test if driver supports append mode.
+
+    Some driver only allow a specific schema. These drivers can be excluded by adding them to blacklist_append_drivers.
+
+    If this test fails, it should be considered to update driver_mode_mingdal in drvsupport.py.
+
+    """
+
+    if driver == "BNA" and GDALVersion.runtime() < GDALVersion(2, 0):
+        # BNA driver segfaults with gdal 1.11
+        return
+
+    path = str(tmpdir.join(get_temp_filename(driver)))
+
+    # If driver is not able to write, we cannot test append
+    if driver in driver_mode_mingdal['w'] and GDALVersion.runtime() < GDALVersion(
+            *driver_mode_mingdal['w'][driver][:2]):
+        return
+
+    # Create test file to append to
+    with fiona.open(path, 'w',
+                    driver=driver,
+                    schema=get_schema(driver)) as c:
+
+        c.write(get_record1(driver))
+
+    if driver in driver_mode_mingdal['a'] and GDALVersion.runtime() < GDALVersion(
+            *driver_mode_mingdal['a'][driver][:2]):
+        # Test if driver really can't append for gdal < driver_mode_mingdal
+        min_version_backup = driver_mode_mingdal['a'][driver]
+        driver_mode_mingdal['a'].pop(driver)
+
+        with pytest.raises(Exception):
+            with fiona.open(path, 'a',
+                            driver=driver) as c:
+                c.write(get_record2(driver))
+
+            with fiona.open(path) as c:
+                assert c.driver == driver
+                assert len([f for f in c]) == 2
+
+        driver_mode_mingdal['a'] = min_version_backup
+
+
+@pytest.mark.parametrize('driver', [driver for driver, raw in supported_drivers.items() if
+                                    raw == 'r' and driver not in blacklist_write_drivers])
+def test_no_write_driver_cannot_write(tmpdir, driver):
+    """Test if read only driver cannot write
+    
+    If this test fails, it should be considered to enable write support for the respective driver in drvsupport.py. 
+    
+    """
+
+    if driver == "BNA" and GDALVersion.runtime() < GDALVersion(2, 0):
+        # BNA driver segfaults with gdal 1.11
+        return
+
+    backup_mode = supported_drivers[driver]
+    supported_drivers[driver] = 'rw'
+
+    path = str(tmpdir.join(get_temp_filename(driver)))
+
+    with pytest.raises(Exception):
+        with fiona.open(path, 'w',
+                        driver=driver,
+                        schema=get_schema(driver)) as c:
+            c.write(get_record1(driver))
+
+    supported_drivers[driver] = backup_mode
```

Comment:
This line is not executed if pytest.raises(Exception) fails. Is there a simple way to setup and teardown the state of supported_drivers?

Reply:
@rbuffat how about using pytest's [monkeypatch fixture](https://docs.pytest.org/en/latest/monkeypatch.html)? I think that will let you do 

`monkeypatch.setitem(fiona.drvsupport.supported_drivers, driver, 'rw')`

once at the very start of the function. I think it would require one import change in the test module.

Reply:
I didn't know about monkeypatch. This looks really handy!